internal/build/docker.go

```diff
@@ -0,0 +1,104 @@
+package build
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"io"
+	"regexp"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+)
+
+type Mount struct {
+	Repo          LocalGithubRepo
+	ContainerPath string
+}
+
+type Repo interface{}
+
+type LocalGithubRepo struct {
+	LocalPath string
+}
+
+type Cmd struct {
+	argv []string
+}
+
+type DockerBuilder interface {
+	BuildDocker(ctx context.Context, baseDockerfile string, mounts []Mount, cmds []Cmd, tag string) (string, error)
+}
+
+type localDockerBuilder struct {
+	dcli *client.Client
+}
+
+// NOTE(dmiller): not fully implemented yet
+func (l *localDockerBuilder) BuildDocker(ctx context.Context, baseDockerfile string, mounts []Mount, cmds []Cmd, tag string) (string, error) {
+	baseTag, err := l.buildBase(ctx, baseDockerfile, tag)
+	if err != nil {
+		return "", err
+	}
+
+	// TODO(dmiller): mounts
+	// TODO(dmiller): steps
+
+	return baseTag, nil
+}
+
+func (l *localDockerBuilder) buildBase(ctx context.Context, baseDockerfile string, tag string) (string, error) {
+	tar, err := tarFromDockerfile(baseDockerfile)
+	if err != nil {
+		return "", err
+	}
+	imageBuildResponse, err := l.dcli.ImageBuild(
+		ctx,
+		tar,
+		types.ImageBuildOptions{
+			Context:    tar,
+			Dockerfile: "Dockerfile",
+			Tags:       []string{tag},
+		})
+
+	defer imageBuildResponse.Body.Close()
+	output := &strings.Builder{}
+	_, err = io.Copy(output, imageBuildResponse.Body)
+	if err != nil {
+		return "", err
+	}
+
+	return getTagFromOutput(output.String()), nil
+}
+
+func tarFromDockerfile(df string) (*bytes.Reader, error) {
+	buf := new(bytes.Buffer)
+	tw := tar.NewWriter(buf)
+	defer tw.Close()
+
+	tarHeader := &tar.Header{
+		Name: "Dockerfile",
+		Size: int64(len(df)),
```

Comment:
usually won't make a difference but technically we want `len([]byte(df))`, no?

Reply:
ooh, good catch. And it turns out that `len(myString)` is returning the number of bytes and not the number of characters or runes. So this is fine!

https://stackoverflow.com/questions/12668681/how-to-get-the-number-of-characters-in-a-string

Reply:
OMG WHAAAAT?!?

https://media0.giphy.com/media/26ufdipQqU2lhNA4g/giphy.gif